File: js/services/baidu.js
/**
 * 百度翻译 API 服务
 * 提供与百度翻译 API 的交互功能
 */

import { logger } from '../utils/logger.js';
import { MD5 } from "../utils/md5.js";

class BaiduTranslateService {
    // API 配置
    static API_CONFIG = {
        BASE_URL: 'https://fanyi-api.baidu.com/api/trans/vip/translate',
        FROM_LANG: 'auto', // 源语言，auto为自动检测
        TO_LANG: 'zh', // 目标语言，默认翻译为中文
    };

    // 错误码映射
    static ERROR_CODES = {
        '52001': '请求超时，请重试',
        '52002': '系统错误，请重试',
        '52003': '未授权用户，请检查appid是否正确或服务是否开通',
        '54000': '必填参数为空，请检查是否少传参数',
        '54001': '签名错误，请检查您的签名生成方法',
        '54003': '访问频率受限，请降低您的调用频率，或进行身份认证后切换为高级版/尊享版',
        '54004': '账户余额不足，请前往管理控制台充值',
        '54005': '长query请求频繁，请降低长query的发送频率，3s后再试',
        '58000': '客户端IP非法，检查个人资料里填写的IP地址是否正确，可前往开发者信息-基本信息修改',
        '58001': '译文语言方向不支持，检查译文语言是否在语言列表里',
        '58002': '服务当前已关闭，请前往百度管理控制台开启服务',
        '58003': '此IP已被封禁',
        '90107': '认证未通过或未生效，请前往我的认证查看认证进度',
        '20003': '请求内容存在安全风险',
    };

    /**
     * 获取错误信息
     */
    static getErrorMessage(code) {
        return this.ERROR_CODES[code] || `未知错误(错误码:${code})`;
    }

    /**
     * 获取配置信息
     */
    static async getConfig() {
        try {
            const response = await fetch('/prompt_assistant/api/config/baidu_translate');
            if (!response.ok) {
                throw new Error('获取配置失败');
            }
            const config = await response.json();
            return config;
        } catch (error) {
            logger.error(`获取百度翻译配置失败: ${error.message}`);
            throw error;
        }
    }

    /**
     * 生成签名
     */
    static async generateSign(query, salt) {
        try {
            const config = await this.getConfig();
            // 拼接字符串
            const str = config.app_id + query + salt + config.secret_key;
            // 计算 MD5
            return MD5(str);
        } catch (error) {
            logger.error(`签名生成失败: ${error.message}`);
            throw error;
        }
    }

    /**
     * 处理长文本 - 按段落分割文本
     * 解决JSONP请求时长文本被截断的问题
     */
    static splitTextByParagraphs(text, maxLength = 2000) {
        if (!text) return [];

        // 按段落分割
        const paragraphs = text.split(/\n+/);
        const chunks = [];
        let currentChunk = "";

        for (const paragraph of paragraphs) {
            // 如果当前段落本身超过最大长度，需要再次分割
            if (paragraph.length > maxLength) {
                // 如果currentChunk不为空，先添加到chunks
                if (currentChunk) {
                    chunks.push(currentChunk);
                    currentChunk = "";
                }

                // 分割长段落
                let remainingText = paragraph;
                while (remainingText.length > 0) {
                    const chunkText = remainingText.slice(0, maxLength);
                    chunks.push(chunkText);
                    remainingText = remainingText.slice(maxLength);
                }
            }
            // 如果添加当前段落会超出长度限制，先保存当前chunk
            else if (currentChunk && (currentChunk.length + paragraph.length + 1 > maxLength)) {
                chunks.push(currentChunk);
                currentChunk = paragraph;
            }
            // 否则，添加到当前chunk
            else {
                if (currentChunk) {
                    currentChunk += "\n" + paragraph;
                } else {
                    currentChunk = paragraph;
                }
            }
        }

        // 添加最后一个chunk
        if (currentChunk) {
            chunks.push(currentChunk);
        }

        return chunks;
    }

    /**
     * 使用 JSONP 方式发送请求
     */
    static jsonp(params) {
        return new Promise((resolve, reject) => {
            // 创建唯一的回调函数名
            const callbackName = 'baiduTranslateCallback_' + Date.now();

            // 创建script标签
            const script = document.createElement('script');

            // 构建URL
            const queryString = Object.keys(params)
                .map(key => `${key}=${encodeURIComponent(params[key])}`)
                .join('&');

            // 设置全局回调函数
            window[callbackName] = (response) => {
                // 清理工作
                delete window[callbackName];
                document.body.removeChild(script);

                if (response.error_code) {
                    const errorMessage = this.getErrorMessage(response.error_code);
                    reject(new Error(`错误：${errorMessage}`));
                } else {
                    resolve(response);
                }
            };

            // 设置script标签的src
            script.src = `${this.API_CONFIG.BASE_URL}?${queryString}&callback=${callbackName}`;

            // 处理加载错误
            script.onerror = () => {
                delete window[callbackName];
                document.body.removeChild(script);
                reject(new Error('错误：网络请求失败'));
            };

            // 添加script标签到页面
            document.body.appendChild(script);
        });
    }

    /**
     * 发送翻译请求
     */
    static async translate(text, from = this.API_CONFIG.FROM_LANG, to = this.API_CONFIG.TO_LANG, request_id = null) {
        // 使用外部传入的request_id，如果没有则自动生成
        request_id = request_id || `baidu_trans_${Date.now()}_${Math.random().toString(36).substr(2, 6)}`;

        try {
            // 参数验证
            if (!text || text.trim() === '') {
                throw new Error('错误：待翻译文本不能为空');
            }

            // 获取配置
            let config;
            try {
                config = await this.getConfig();
                if (!config || !config.app_id || !config.secret_key) {
                    throw new Error('配置无效');
                }
            } catch (error) {
                logger.error(`获取配置失败: ${error.message}`);
                throw new Error('错误：请先配置百度翻译 API 的 APP_ID 和 SECRET_KEY');
            }

            // 记录请求开始
            logger.debug(`发起百度API请求 | 请求ID:${request_id} | API:baidu_translate | 参数:${JSON.stringify({ text, from, to })}`);

            // 处理长文本 - 按段落分割
            const textChunks = this.splitTextByParagraphs(text);
            let translatedText = '';

            // 如果文本被分割成多个段落，依次翻译
            if (textChunks.length > 1) {
                logger.debug(`文本分段处理 | 请求ID:${request_id} | 总段数:${textChunks.length}`);

                for (let i = 0; i < textChunks.length; i++) {
                    const chunk = textChunks[i];
                    const salt = Date.now();
                    const sign = await this.generateSign(chunk, salt);

                    const params = {
                        q: chunk,
                        from,
                        to,
                        appid: config.app_id,
                        salt: salt,
                        sign: sign
                    };

                    const response = await this.jsonp(params);
                    translatedText += response.trans_result[0].dst + (i < textChunks.length - 1 ? '\n' : '');

                    // 如果还有更多段落，等待一段时间再继续
                    if (i < textChunks.length - 1) {
                        await new Promise(resolve => setTimeout(resolve, 1000));
                    }
                }
            } else {
                const chunk = textChunks[0] || text;
                const salt = Date.now();
                const sign = await this.generateSign(chunk, salt);

                const params = {
                    q: chunk,
                    from,
                    to,
                    appid: config.app_id,
                    salt: salt,
                    sign: sign
                };

                const response = await this.jsonp(params);
                translatedText = response.trans_result[0].dst;
            }

            // 记录请求完成
            logger.debug(`百度API请求完成 | 请求ID:${request_id}`);

            // 返回统一格式的结果
            return {
                success: true,
                data: {
                    translated: translatedText,
                    from: from,
                    to: to,
                    original: text
                }
            };

        } catch (error) {
            logger.error(`百度API请求失败 | 请求ID:${request_id} | 错误:${error.message}`);
            return {
                success: false,
                error: error.message
            };
        }
    }

    /**
     * 批量翻译
     */
    static async batchTranslate(texts, from = this.API_CONFIG.FROM_LANG, to = this.API_CONFIG.TO_LANG) {
        try {
            if (!Array.isArray(texts) || texts.length === 0) {
                throw new Error('待翻译文本数组不能为空');
            }

            // 串行处理每个文本的翻译
            const results = [];
            for (const text of texts) {
                const result = await this.translate(text, from, to);
                results.push(result);
            }

            return results;

        } catch (error) {
            logger.error(`批量翻译 | 结果:失败 | 错误:${error.message}`);
            return [];
        }
    }

    /**
     * 设置 API 配置
     */
    static setConfig(config) {
        try {
            // 验证必要的配置项
            if (config.app_id && config.secret_key) {
                this.API_CONFIG = {
                    ...this.API_CONFIG,
                    app_id: config.app_id,
                    secret_key: config.secret_key
                };
                logger.debug('API配置更新成功');
                return true;
            } else {
                throw new Error('app_id 和 secret_key 是必需的配置项');
            }
        } catch (error) {
            logger.error(`API配置更新失败 | 错误:${error.message}`);
            return false;
        }
    }

    /**
     * 获取当前 API 配置
     */
    static getAPIConfig() {
        return { ...this.API_CONFIG };
    }
}

export { BaiduTranslateService };

File: services/baidu.py
import random
import httpx
from hashlib import md5
import json
import time
import re
from typing import Optional, Dict, Any, List
import asyncio
from ..utils.common import BAIDU_ERROR_CODE_MESSAGES, ProgressBar, log_complete, log_error, TASK_TRANSLATE
from .core import HTTPClientPool

class BaiduTranslateService:
    @staticmethod
    def split_text_by_paragraphs(text, max_length=2000):
        """
        按段落分割文本，处理长文本翻译
        使用正则表达式匹配连续的换行符，与JavaScript版本保持一致
        """
        if not text:
            return []

        # 保留原始的换行符，用于后续恢复格式
        # 先将文本按换行符分割为行
        lines = text.split('\n')
        chunks = []
        current_chunk = ""
        
        for line in lines:
            # 如果当前行本身超过最大长度，需要再次分割
            if len(line) > max_length:
                # 如果current_chunk不为空，先添加到chunks
                if current_chunk:
                    chunks.append(current_chunk)
                    current_chunk = ""
                
                # 分割长行
                remaining_text = line
                while len(remaining_text) > 0:
                    chunk_text = remaining_text[:max_length]
                    chunks.append(chunk_text)
                    remaining_text = remaining_text[max_length:]
            # 如果添加当前行会超出长度限制，先保存当前chunk
            elif current_chunk and (len(current_chunk) + len(line) + 1 > max_length):
                chunks.append(current_chunk)
                current_chunk = line
            # 否则，添加到当前chunk
            else:
                if current_chunk:
                    current_chunk += "\n" + line
                else:
                    current_chunk = line
        
        # 添加最后一个chunk
        if current_chunk:
            chunks.append(current_chunk)
        
        return chunks

    @staticmethod
    async def translate_chunk(client, chunk, app_id, secret_key, from_lang, to_lang, retry_count=1):
        """
        异步翻译单个文本块，带重试机制
        """
        for attempt in range(retry_count):
            try:
                # 生成签名
                salt = random.randint(32768, 65536)
                sign = md5((app_id + chunk + str(salt) + secret_key).encode('utf-8')).hexdigest()
                
                # 构建请求
                url = 'https://fanyi-api.baidu.com/api/trans/vip/translate'
                params = {
                    'q': chunk,
                    'from': from_lang,
                    'to': to_lang,
                    'appid': app_id,
                    'salt': salt,
                    'sign': sign
                }
                
                # 发送异步请求
                response = await client.post(url, data=params)
                
                if response.status_code != 200:
                    if attempt < retry_count - 1:
                        await asyncio.sleep(1)
                        continue
                    raise Exception(f"百度: HTTP请求失败，状态码: {response.status_code}")

                result = response.json()
            
                # 检查错误
                if 'error_code' in result:
                    error_code = result['error_code']
                    error_message = BAIDU_ERROR_CODE_MESSAGES.get(
                        error_code, 
                        f"未知错误(错误码:{error_code})"
                    )
                    
                    # 某些错误码可以重试
                    if error_code in ['54003', '52001', '52002'] and attempt < retry_count - 1:
                        await asyncio.sleep(1)
                        continue
                        
                    raise Exception(f"百度: {error_message}")

                # 处理翻译结果
                if 'trans_result' in result and result['trans_result']:
                    translated_parts = [item['dst'] for item in result['trans_result']]
                    return '\n'.join(translated_parts)
                else:
                    raise Exception("百度: 翻译结果为空")
                    
            except (httpx.HTTPError, asyncio.TimeoutError) as e:
                if attempt < retry_count - 1:
                    from ..utils.common import WARN_PREFIX
                    print(f"\r{WARN_PREFIX} 百度翻译请求遇到网络错误，尝试重试 ({attempt+1}/{retry_count}): {e}")
                    await asyncio.sleep(1)
                else:
                    raise Exception(f"百度: 网络请求失败，请检查网络连接或稍后再试 ({type(e).__name__})")
            except Exception as e:
                if attempt < retry_count - 1:
                    await asyncio.sleep(1)
                else:
                    if str(e).startswith("百度:"):
                        raise e
                    else:
                        raise Exception(f"百度: 翻译过程中发生未知错误 ({type(e).__name__})")
                    
        raise Exception("百度: 超过最大重试次数")

    @staticmethod
    async def translate(text, from_lang='auto', to_lang='zh', request_id=None, is_auto=False, cancel_event=None, task_type=None, source=None):
        """
        异步调用百度翻译API进行翻译
        
        参数:
            task_type: 任务类型，用于统一日志输出
            cancel_event: 中断事件(保持接口一致性,百度翻译暂不支持中断)
        """
        try:
            request_id = request_id or f"baidu_trans_{int(time.time())}_{random.randint(1000, 9999)}"
            
            if not text or text.strip() == '':
                return {"success": False, "error": "百度: 待翻译文本不能为空"}
            
            from ..config_manager import config_manager
            config = config_manager.get_baidu_translate_config()
            
            app_id = config.get('app_id')
            secret_key = config.get('secret_key')
            
            if not app_id or not secret_key:
                return {"success": False, "error": "百度: 请先配置百度翻译API的APP_ID和SECRET_KEY"}

            # 确定任务类型
            task_type = task_type or TASK_TRANSLATE

            text_chunks = BaiduTranslateService.split_text_by_paragraphs(text)
            if not text_chunks:
                text_chunks = [text]

            translated_parts = []
            
            # 获取HTTP客户端（复用连接池，不使用代理）
            # 注意：百度API通常在国内直连更快，所以不配置proxy参数
            client = HTTPClientPool.get_client(
                provider="baidu_translate",
                base_url="https://fanyi-api.baidu.com",
                timeout=10.0
            )
            
            # 由于HTTPClientPool.get_client返回的是AsyncClient实例，我们可以直接使用
            # 但为了确保不使用系统代理（模拟 trust_env=False），我们需要在get_client中支持或在这里手动创建
            # 考虑到HTTPClientPool目前的设计，我们直接手动创建一个不使用代理的客户端可能更稳妥
            # 或者修改HTTPClientPool支持 trust_env=False。
            # 为了保持统一，我们使用HTTPClientPool，但如果需要强制直连，我们可以传递特定的代理配置
            
            # 实际上，HTTPClientPool.get_client 如果没有传proxy，会使用系统代理。
            # 百度翻译之前的实现是 trust_env=False (禁用系统代理)。
            # 为了保持一致，我们这里手动创建一个专用的客户端，或者修改HTTPClientPool。
            # 考虑到百度翻译的特殊性（国内直连），手动创建一个简单的客户端是最安全的迁移方式。
            
            # 创建统一进度条
            from ..server import is_streaming_progress_enabled
            pbar = ProgressBar(
                request_id=request_id,
                service_name="百度翻译",
                streaming=is_streaming_progress_enabled(),
                extra_info=f"长度:{len(text)}",
                task_type=task_type,
                source=source
            )
            
            start_time = time.perf_counter()
            
            async with httpx.AsyncClient(trust_env=False, verify=False, timeout=10.0) as client:
                for i, chunk in enumerate(text_chunks):
                    # ---中断监控---
                    is_interrupted = False
                    if cancel_event is not None and cancel_event.is_set():
                        is_interrupted = True
                    else:
                        try:
                            from server import PromptServer
                            if hasattr(PromptServer.instance, 'execution_interrupted') and PromptServer.instance.execution_interrupted:
                                is_interrupted = True
                        except: pass
                    
                    if is_interrupted:
                        pbar.cancel(f"{WARN_PREFIX} 任务被中断 | 服务:百度翻译")
                        return {"success": False, "error": "任务被中断", "interrupted": True}
                    # ------------

                    try:
                        chunk_translation = await BaiduTranslateService.translate_chunk(
                            client, chunk, app_id, secret_key, from_lang, to_lang
                        )
                        translated_parts.append(chunk_translation)

                        if i < len(text_chunks) - 1:
                            await asyncio.sleep(1)

                    except Exception as chunk_error:
                        # 输出错误日志
                        pbar.error(str(chunk_error))
                        return {"success": False, "error": str(chunk_error)}
            
            translated_text = '\n'.join(translated_parts)
            # 完成阶段
            elapsed = int((time.perf_counter() - start_time) * 1000)
            pbar.done(char_count=len(translated_text), elapsed_ms=elapsed)

            return {
                "success": True,
                "data": {
                    "translated": translated_text,
                    "from": from_lang,
                    "to": to_lang,
                    "original": text
                }
            }
            
        # 关键修复：单独捕获外层 CancelledError，确保 pbar 被正确停止
        except asyncio.CancelledError:
            if 'pbar' in locals() and pbar:
                pbar.cancel(f"{WARN_PREFIX} 任务被外部取消 | 服务:百度翻译")
            return {"success": False, "error": "任务被取消", "interrupted": True}
        
        except Exception as e:
            # 输出错误日志
            if 'pbar' in locals() and pbar:
                pbar.error(str(e))
            return {"success": False, "error": str(e)}
    
    @staticmethod
    async def batch_translate(texts, from_lang='auto', to_lang='zh'):
        """
        异步批量翻译文本
        """
        tasks = [BaiduTranslateService.translate(text, from_lang, to_lang) for text in texts]
        return await asyncio.gather(*tasks)